authz/providers/azure/azure.go

```diff
@@ -0,0 +1,150 @@
+	authz.SupportedOrgs = append(authz.SupportedOrgs, OrgType)
+}
+
+type Authorizer struct {
+	rbacClient *rbac.AccessInfo
+}
+
+func New(opts authzOpts.Options, authopts auth.Options) (authz.Interface, error) {
+	once.Do(func() {
+		glog.Info("Creating Azure global authz client")
+		client, err = newAuthzClient(opts, authopts)
+		if client == nil || err != nil {
+			glog.Fatalf("Authz RBAC client creation failed. Error: %s", err)
+		}
+	})
+	return client, err
+}
+
+func newAuthzClient(opts authzOpts.Options, authopts auth.Options) (authz.Interface, error) {
+	c := &Authorizer{}
+
+	authzInfoVal, err := getAuthzInfo(authopts.Environment)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error in getAuthzInfo %s")
+	}
+
+	c.rbacClient, err = rbac.New(opts, authopts, authzInfoVal)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create ms rbac client")
+	}
+
+	return c, nil
+}
+
+func (s Authorizer) Check(request *authzv1.SubjectAccessReviewSpec, store authz.Store) (*authzv1.SubjectAccessReviewStatus, error) {
+	if request == nil {
+		return nil, errors.New("subject access review is nil")
+	}
+
+	// check if user is system accounts
+	if strings.HasPrefix(strings.ToLower(request.User), "system") {
+		glog.V(3).Infof("returning no op to system accounts")
+		return &authzv1.SubjectAccessReviewStatus{Allowed: false, Reason: rbac.NoOpinionVerdict}, nil
+	}
+
+	if s.rbacClient.SkipAuthzCheck(request) {
+		glog.V(3).Infof("user %s is part of skip authz list. returning no op.", request.User)
+		return &authzv1.SubjectAccessReviewStatus{Allowed: false, Reason: rbac.NoOpinionVerdict}, nil
+	}
+
+	if _, ok := request.Extra["oid"]; !ok {
+		if s.rbacClient.ShouldSkipAuthzCheckForNonAADUsers() {
+			glog.V(3).Infof("Skip RBAC is set for non AAD users. Returning no opinion for user %s. You may observe this for AAD users for 'can-i' requests.", request.User)
+			return &authzv1.SubjectAccessReviewStatus{Allowed: false, Reason: rbac.NoOpinionVerdict}, nil
+		} else {
+			glog.V(3).Infof("Skip RBAC for non AAD user is not set. Returning deny access for non AAD user %s. You may observe this for AAD users for 'can-i' requests.", request.User)
+			return &authzv1.SubjectAccessReviewStatus{Allowed: false, Denied: true, Reason: rbac.NotAllowedForNonAADUsers}, nil
+		}
+	}
+
+	exist, result := s.rbacClient.GetResultFromCache(request, store)
+	if exist {
+		if result {
+			glog.V(3).Infof("cache hit: returning allowed to user %s", request.User)
+			return &authzv1.SubjectAccessReviewStatus{Allowed: result, Reason: rbac.AccessAllowedVerdict}, nil
+		} else {
+			glog.V(3).Infof("cache hit: returning denied to user %s", request.User)
+			return &authzv1.SubjectAccessReviewStatus{Allowed: result, Denied: true, Reason: rbac.AccessNotAllowedVerdict}, nil
+		}
+	}
+
+	// if set true, webhook will allow access to discovery APIs for authenticated users. If false, access check will be performed on Azure.
+	if s.rbacClient.AllowNonResPathDiscoveryAccess(request) {
```

Comment:
Is this necessary because Kubernetes automatically creates `system:discovery` cluster role and cluster role binding to give this permission to any authenticated users.

Reply:
webhook will be enabled before RBAC and we need to handle this. As Kubernetes automatically adds all authenticated users to system:discovery, we are good to allow access on these paths. Hence I have added this check which would be true by default.